### PR TITLE
New `run-tests.ts`, making test runs more reliable

### LINKED ISSUE #1400: Make tests pass on the updated testing infrastructure
#1389 improves the testing infrastructure and it seems that with enough "luck", by which I mean a good order of tests or their combination, all tests even pass. However, in practice, it's not very repeatable or reliable.

This issue is about getting the tests run reasonably reliably on both Mac and Windows. There are still debt items mentioned in #1284 like the "temporarily" disabled [`ActivationDeactivationTest`](https://github.com/versionpress/versionpress/blob/e17f0602ffe49679517a2921e4edf34cfe9b1c46/plugins/versionpress/tests/Selenium/ActivationDeactivationTest.php#L14) and [`ThemeCustomizerTest`](https://github.com/versionpress/versionpress/blob/f4a00328d7903ac1c7bcc10673ce5dc621e23e9f/plugins/versionpress/tests/Selenium/ThemeCustomizerTest.php) which are out of scope of this issue.

- [x] Review Docker stack startup & waiting – @pavelevap has reported that some of the tests are reporting errors (`E`, not `F`) as for example database is not ready yet.
    - Resolved by `run-tests.ts`, #1401.
- [x] Make `npm run tests` pass.
    - With improved test runner #1401 and a workaround for `cloneLooksExactlySameAsOriginal` #1410, all tests now pass. 🎉

### PR DESCRIPTION
**Pull request by borekb**
Issues: #1389 & #1400

The main goal of this PR is to improve reliability of our test runs – `npm run tests:full` should produce consistent results across operating systems and several runs on the same machine.

Highlights:

- New `scripts/run-tests.ts` instead of direct invocation of docker-compose and PHPUnit. This allows better control over how the tests run, which containers are started and in which order, there's proper waiting between them (c1a1fbb7021e6675ed46976aa4bc74c095eba446), etc. This resolves many confusing behaviors when the computer was too slow to start or something similar.
- The script is smart enough to start or not to start WordPress depending on the test suite being run.
- The whole Docker stack is now cleared before tests start. Previous runs cannot influence the current one.

Full TODO list:

- [x] Add `scripts/run-tests.ts` to orchestrate test runs.
- [x] Clear whole Docker stack before starting tests.
- [x] Imply WP container startup from specific test suites. For example, `--testsuite Workflow` should start WP.
- [x] Remove debugging scripts and info, reintroduce it later (we have some todo items related to Xdebug in #1389).
- [x] Add an argument to provide custom phpunit.xml
- [x] Remove `tests:custom` scripts from `package.json`.
- [x] Update docs in `testing.md`.
- ~~Fail immediately by default, flag to let them run~~ I've decided against it for now.
- ~~On Ctrl+C to the Node.js script, gracefully shut down containers.~~ Postponed; the user has to clean up manually.
- [x] Properly colorize PHPUnit output.
- [x] Pass through standard PHPUnit arguments
- [x] Simplify calling `node -r ts-node/register run-tests ...` to `npm run tests -- ...`.

Noted for followup PRs:

- Inspecting site via `docker-compose -f docker-compose-tests.yml up -d wordpress-for-tests mysql-for-tests` is a bit weird, review it.
- Should we still have `tests:cleanup`?

### CODE REVIEW

**Comment by borekb** [comment]
1778c7f shows how moving Workflow tests above End2End tests in `phpunit.xml` fixes https://github.com/versionpress/versionpress/issues/1400#issuecomment-471987716 for me.

However, it also makes tests run considerably slower – 30 minutes instead of 10 for `npm run tests:full`.

I think we should clean up the testing environment before each "big" test suite.

**Comment by borekb** [comment]
https://github.com/versionpress/versionpress/pull/1401/commits/c1a1fbb7021e6675ed46976aa4bc74c095eba446 deals with another problem that @pavelevap frequently encounters – tests run before the Docker containers are fully started (WP + MySQL take around 20 seconds to start on my machine, can be longer depending on a specific device).

The commit shows an approach where a waiting container is added to `docker-compose` and a test running script uses it to delay the test execution. Key parts:

https://github.com/versionpress/versionpress/blob/c1a1fbb7021e6675ed46976aa4bc74c095eba446/docker-compose-tests.yml#L59-L62

https://github.com/versionpress/versionpress/blob/c1a1fbb7021e6675ed46976aa4bc74c095eba446/scripts/run-tests.ts#L15

When I comment out the waiting line in `run-tests.ts`, I get this error:

```
Exception: Error executing cmd 'php '/tmp/wp-cli-latest-stable.phar' db query --dbuser='root' --dbpass='r00tpwd' 'SELECT User FROM mysql.user'' from working directory '/var/www/html/wptest':
ERROR 2002 (HY000): Can't connect to MySQL server on 'mysql' (115)
```

With the waiting in place, tests pass fine.

Note that on my machine, to make the test fail, it was necessary to start the MySQL from an entirely clean state – with the volume deleted. That's why the `run-tests.ts` script first calls `docker-compose down -v`.

**Comment by borekb** [comment]
The [recent commits](https://github.com/versionpress/versionpress/pull/1401#commits-pushed-0f4439d) implemented `run-tests.ts` to a reasonable state where it can run both unit and all tests better than the raw npm scripts before:

Screenshot 2019-03-17 at 22 38 06

Screenshot 2019-03-17 at 22 37 00

The script has this "API" for now:

```
  Usage
    $ run-tests ...

  Options
    --testsuite         Testsuite from phpunit.xml to run. Can be repeated
                        multiple times.
    --with-wordpress    Start WordPress & MySQL containers. WIP: is implied
                        from --testsuite in some cases.
    -h, --help          Show help.

  Examples
    $ run-tests --testsuite Unit
```

TODO:

- Imply `--with-wordpress` from specific test suites. For example, `--testsuite Workflow` should start WP.
- Document how `run-tests.ts` relates to `phpunit.xml`.
- Add an argument to provide custom phpunit.xml – this will allow us to remove `tests:custom` scripts from `package.json`.
- Update public docs in `testing.md`.

**Comment by borekb** [comment]
BTW I'm pretty happy with the test reliability now:

- The tests wait for containers to start.
- I get consistent results on both Mac and Docker Toolbox on Windows (documented in #1403).
- The one failing test – https://github.com/versionpress/versionpress/issues/1400#issuecomment-471987716 – is probably a real bug, @pavelevap will be creating an issue for it.

The `run-tests.ts` scripts also provides foundation for future orchestration of tests outside of PHPUnit. For example, we'll be able to run full tests with WP-CLI worker and then with Selenium worker quite easily – this TypeScript script will simply run PHPUnit twice. Also, something like `--clean-wp-before-each-testsuite` can be added quite easily, I like how the test runs will be more customizable.

File: docker-compose-tests.yml [hunk 1]
```diff
@@ -1,16 +1,19 @@
 version: '3.6'
 services:
 
+  # !!! Containers in this file are meant to be started via `scripts/run-tests.ts`, not manully.
```
Comment (pavelevap) [inline docker-compose-tests.yml]: Small typo "manully".

Comment (borekb) [reply docker-compose-tests.yml]: Thanks, resolved in https://github.com/versionpress/versionpress/pull/1401/commits/8669fa4654aab8c19e9d00eee510ff98649f7dc9.

**Comment by borekb** [comment]
@pavelevap The PR is now ready for a review. Docs preview [here](https://deploy-preview-1401--vp-docs.netlify.com/en/developer/testing/).

**Comment by borekb** [comment]
I'm getting consistent results between several machines now. 🎉

### MacBook Pro with Docker Desktop

MacBook Pro (2015)\
macOS Mochave 10.14.3\
Intel Core i7 @ 2.5 GHz, 4 Cores, 8 Logical Processors\
16 GB RAM\
Docker Desktop for Mac 2.0.0.3 (build 31259, engine 18.09.2)

Tests took **11 minutes**.

Screenshot 2019-03-24 at 20 49 51

### Windows laptop with Docker Desktop

Dell XPS 13 9350 (2015)\
Windows 10 v1809 ("October 2018 Update")\
Core i5-6200U CPU @ 2.30GHz, 2 Cores, 4 Logical Processors\
8 GB RAM\
Docker Desktop for Windows 2.0.0.3 (build 8858db3, engine 18.09.2)

Tests took **18 minutes**. (The CLI environment is using [this MSYS2 setup](https://medium.com/@borekb/zsh-via-msys2-on-windows-3964a943b1ce).)

image

### Windows laptop with Docker Toolbox

HP ProBook 650 G1 (2014)\
Windows 10 v1809 ("October 2018 Update")\
Core i7-4702MQ CPU @ 2.20GHz, 4 Core(s), 8 Logical Processors\
8 GB RAM\
Docker Toolbox [v18.09.3](https://github.com/docker/toolbox/releases/tag/v18.09.3)

Tests took **37 minutes**.

image

I tried with both the default VM configuration (1GB RAM, 1 CPU) and an [upgraded VM](https://github.com/crops/docker-win-mac-docs/wiki/Windows-Instructions-(Docker-Toolbox)) with 4GB RAM and 2 CPU but it didn't have any effect, the tests still took about 37 minutes. It's probably limited by the virtualization technology; the tests cannot utilize the more resources:

image

### Test output

On all machines, the console output was this:

```
$ npm run tests

> versionpress-dev-workspace@1.0.0 tests /Users/borekb/dev/VersionPress/versionpress
> node -r ./scripts/node_modules/ts-node/register scripts/run-tests.ts


> Cleaning up Docker containers and volumes...
Removing network versionpress_default
Network versionpress_default not found.
Removing volume versionpress_db-data-for-tests
Removing volume versionpress_wpcli-cache
Removing volume versionpress_wordpress-files
Removing volume versionpress_test-logs

> Starting MySQL...
Creating network "versionpress_default" with the default driver
Creating volume "versionpress_db-data-for-tests" with default driver
Creating volume "versionpress_wpcli-cache" with default driver
Creating volume "versionpress_wordpress-files" with default driver
Creating volume "versionpress_test-logs" with default driver
Creating versionpress_mysql-for-tests_1 ... done
Waiting for mysql-for-tests:3306  .........  up!
Everything is up

> Starting WordPress...
Creating versionpress_wordpress-for-tests_1 ... done
Waiting for wordpress-for-tests:80  ..  up!
Everything is up

> Running tests...
PHPUnit 5.7.20 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.2.16 with Xdebug 2.6.0
Configuration: /opt/versionpress/tests/phpunit.xml

...............................................................  63 / 738 (  8%)
............................................................... 126 / 738 ( 17%)
............................................................... 189 / 738 ( 25%)
............................................................... 252 / 738 ( 34%)
............................................................... 315 / 738 ( 42%)
............................................................... 378 / 738 ( 51%)
............................................................... 441 / 738 ( 59%)
............................................................... 504 / 738 ( 68%)
............................................................... 567 / 738 ( 76%)
............................................................... 630 / 738 ( 85%)
............S....SS....S..S....S................S....S.S.SSS... 693 / 738 ( 93%)
........S....................S.....FSS......Php Execution Time: 0.032222986221313 Sec
.                   738 / 738 (100%)Bash Execution Time: 0.035927057266235 Sec


Time: 11.13 minutes, Memory: 18.00MB

There was 1 failure:

1) VersionPress\Tests\Workflow\CloneMergeTest::cloneLooksExactlySameAsOriginal
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-		Posted on November 11, 2011Test post for revert
+		Posted on November 11, 2011March 24, 2019Test post
@@ @@

(...see full output below; the diff is several hundreds lines long...)

/opt/versionpress/tests/Workflow/CloneMergeTest.php:210
/opt/versionpress/tests/Workflow/CloneMergeTest.php:48

--

There were 16 skipped tests:

1) VersionPress\Tests\End2End\Media\MediaTest::editationOfFileCreatesCommit
File cannot be changed using WP-CLI.

/opt/versionpress/tests/End2End/Media/MediaTestWpCliWorker.php:48
/opt/versionpress/tests/End2End/Media/MediaTest.php:89

2) VersionPress\Tests\End2End\Menus\MenusTest::creatingMenuItemDraftLeavesNotCleanWorkingDirectory
There is no way to create menu item draft in the WP-CLI

/opt/versionpress/tests/End2End/Menus/MenusWpCliWorker.php:57
/opt/versionpress/tests/End2End/Menus/MenusTest.php:101

3) VersionPress\Tests\End2End\Menus\MenusTest::orphanedMenuItemShouldBeDeletedWhenDoingUndo
This test depends on "VersionPress\Tests\End2End\Menus\MenusTest::creatingMenuItemDraftLeavesNotCleanWorkingDirectory" to pass.

4) VersionPress\Tests\End2End\Options\OptionsTest::changingMoreOptionsCreatesOptionEditAction
More options cannot be changed at once using WP-CLI

/opt/versionpress/tests/End2End/Options/OptionsTestWpCliWorker.php:26
/opt/versionpress/tests/End2End/Options/OptionsTest.php:39

5) VersionPress\Tests\End2End\Pages\PagesTest::updatingPageViaQuickEditWorksEquallyWell
There is nothing like quick edit in the WP-CLI

/opt/versionpress/tests/End2End/Utils/PostTypeTestWpCliWorker.php:51
/opt/versionpress/tests/End2End/Utils/PostTypeTestCase.php:55
/opt/versionpress/tests/End2End/Pages/PagesTest.php:38

6) VersionPress\Tests\End2End\Pages\PagesTest::previewingDraftDoesNotCreateCommit
There is nothing like preview in the WP-CLI

/opt/versionpress/tests/End2End/Utils/PostTypeTestWpCliWorker.php:100
/opt/versionpress/tests/End2End/Utils/PostTypeTestCase.php:127
/opt/versionpress/tests/End2End/Pages/PagesTest.php:93

7) VersionPress\Tests\End2End\Posts\PostsTest::updatingPostViaQuickEditWorksEquallyWell
There is nothing like quick edit in the WP-CLI

/opt/versionpress/tests/End2End/Utils/PostTypeTestWpCliWorker.php:51
/opt/versionpress/tests/End2End/Utils/PostTypeTestCase.php:55
/opt/versionpress/tests/End2End/Posts/PostsTest.php:39

8) VersionPress\Tests\End2End\Posts\PostsTest::previewingDraftDoesNotCreateCommit
There is nothing like preview in the WP-CLI

/opt/versionpress/tests/End2End/Utils/PostTypeTestWpCliWorker.php:100
/opt/versionpress/tests/End2End/Utils/PostTypeTestCase.php:127
/opt/versionpress/tests/End2End/Posts/PostsTest.php:94

9) VersionPress\Tests\End2End\Posts\PostsTest::previewingUnsavedPostCreatesDraft
There is nothing like preview in the WP-CLI

/opt/versionpress/tests/End2End/Utils/PostTypeTestWpCliWorker.php:128
/opt/versionpress/tests/End2End/Utils/PostTypeTestCase.php:145
/opt/versionpress/tests/End2End/Posts/PostsTest.php:115

10) VersionPress\Tests\End2End\Posts\PostsTest::settingFeaturedImageToUnsavedPostDoesNotCommit
Featured image cannot be assigned to unexisting post using WP-CLI

/opt/versionpress/tests/End2End/Utils/PostTypeTestWpCliWorker.php:137
/opt/versionpress/tests/End2End/Utils/PostTypeTestCase.php:175
/opt/versionpress/tests/End2End/Posts/PostsTest.php:144

11) VersionPress\Tests\End2End\Posts\PostsTest::turningUnsavedPostWithFeaturedImageIntoDraftSavesTheFeaturedImage
This test depends on "VersionPress\Tests\End2End\Posts\PostsTest::settingFeaturedImageToUnsavedPostDoesNotCommit" to pass.

12) VersionPress\Tests\End2End\Posts\PostsTest::changingPostFormatUpdatesItsTaxonomy
Post format cannot be changed using WP-CLI.

/opt/versionpress/tests/End2End/Posts/PostsTestWpCliWorker.php:49
/opt/versionpress/tests/End2End/Posts/PostsTest.php:164

13) VersionPress\Tests\End2End\Revert\RevertTest::clickingOnCancelOnlyHidesThePopup
There is no cancel button in the WP-CLI

/opt/versionpress/tests/End2End/Revert/RevertTestWpCliWorker.php:96
/opt/versionpress/tests/End2End/Revert/RevertTest.php:117

14) VersionPress\Tests\End2End\Users\UsersTest::editingMultipleUsermetaCreatesBulkAction
There is no way to change multiple usermeta using WP-CLI

/opt/versionpress/tests/End2End/Users/UsersTestWpCliWorker.php:139
/opt/versionpress/tests/End2End/Users/UsersTest.php:154

15) VersionPress\Tests\Workflow\CloneMergeTest::updatedCloneCanBeMergedBack
This test depends on "VersionPress\Tests\Workflow\CloneMergeTest::cloneLooksExactlySameAsOriginal" to pass.

16) VersionPress\Tests\Workflow\CloneMergeTest::updatedSiteCanBePushedToClone
This test depends on "VersionPress\Tests\Workflow\CloneMergeTest::cloneLooksExactlySameAsOriginal" to pass.

FAILURES!
Tests: 738, Assertions: 950, Failures: 1, Skipped: 16.

> Stopping containers...
Stopping versionpress_wordpress-for-tests_1 ... done
Stopping versionpress_mysql-for-tests_1     ... done
Removing versionpress_wordpress-for-tests_1 ... done
Removing versionpress_mysql-for-tests_1     ... done
Removing network versionpress_default
```

_Note: the above is omits the failing test with many empty lines, see https://github.com/versionpress/versionpress/issues/1400#issuecomment-471987716._

**Pull request merged**